engine/classes/Elgg/I18n/MessageTranslator.php

```diff
@@ -0,0 +1,52 @@
+<?php
+namespace Elgg\I18n;
+
+/**
+ * WARNING: API IN FLUX. DO NOT USE DIRECTLY.
+ *
+ * @access private
+ *
+ * @since 1.11
+ */
+class MessageTranslator implements TranslatorInterface {
+	
+	// TODO Maybe this should be an array of locales to provide multiple fallbacks?
+	/** @var Locale */
+	private $defaultLocale;
+	
+	/** @var MessageBundle */
+	private $messages;
+	
+	/**
+	 * Constructor
+	 * 
+	 * @param MessageBundle $messages      Messages that this translator is aware of
+	 * @param Locale        $defaultLocale The fallback locale
+	 */
+	public function __construct(MessageBundle $messages, Locale $defaultLocale) {
+		$this->defaultLocale = $defaultLocale;
+		$this->messages = $messages;
+	}
+	
+	/** @inheritDoc */
+	function translate($key, array $args = [], Locale $locale = null) {
+		$locales = [
+			$locale,
+			$this->defaultLocale,
+			Locale::parse('en'),
+		];
+		
+		foreach ($locales as $locale) {
+			if (!isset($locale)) {
+				continue;
+			}
+			
+			$message = $this->messages->get($key, $locale);
+			if (isset($message)) {
+				return empty($args) ? "$message" : $message->format($args);
```

Comment:
This was a documented requirement for BC from the days that folks did their own sprintf around elgg_echo.